message/router/middleware/deduplicator.go

```diff
@@ -0,0 +1,237 @@
+		if err != nil && err != io.EOF {
+			return "", err
+		}
+		return string(h.Sum(nil)), nil
+	}
+}
+
+// NewMessageHasherFromMetadataField looks for a hash value
+// inside message metadata instead of calculating a new one.
+// Useful if a [MessageHasher] was applied in a previous
+// [message.HandlerFunc].
+func NewMessageHasherFromMetadataField(field string) MessageHasher {
+	return func(m *message.Message) (string, error) {
+		fromMetadata, ok := m.Metadata[field]
+		if ok {
+			return fromMetadata, nil
+		}
+		return "", fmt.Errorf("cannot recover hash value from metadata of message #%s: field %q is absent", m.UUID, field)
+	}
+}
+
+type Deduplicator struct {
+	tagger MessageHasher
+	window time.Duration
+
+	mu   *sync.Mutex
+	tags map[string]time.Time
+}
+
+func (d *Deduplicator) cleanOutLoop(ctx context.Context, ticker *time.Ticker) {
+	for {
+		select {
+		case <-ctx.Done():
+			return // execution ended, part the go routine
+		case tagsBefore := <-ticker.C:
+			d.cleanOut(tagsBefore)
+		}
+	}
+}
+
+func (d *Deduplicator) cleanOut(tagsBefore time.Time) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	for hash, expires := range d.tags {
+		if expires.Before(tagsBefore) {
+			delete(d.tags, hash)
+		}
+	}
+}
+
+// Len returns the number of known tags that have not been
+// cleaned out yet.
+func (d *Deduplicator) Len() (count int) {
+	d.mu.Lock()
+	count = len(d.tags)
+	d.mu.Unlock()
+	return
+}
+
+// IsDuplicate returns true if the message hash tag calculated
+// using a [MessageHasher] was seen in deduplication time window.
+func (d *Deduplicator) IsDuplicate(m *message.Message) (bool, error) {
+	tag, err := d.tagger(m)
+	if err != nil {
+		return false, err
+	}
+
+	d.mu.Lock()
+	_, alreadySeen := d.tags[tag]
+	if alreadySeen {
+		// NOTE: could also check if tag expires.After(t)
+		// and remove it for exact expiration
+		// instead of fuzzy until-next clean up expiration
+		// but this should not be needed for most use cases.
+		d.mu.Unlock()
+		return true, nil
+	}
+	d.tags[tag] = time.Now().Add(d.window)
+	d.mu.Unlock()
+	return false, nil // first time, not a duplicate
+}
+
+// NewDeduplicator returns a new Deduplicator.
+// Call [Deduplicator.Middleware] for a new middleware
+// or [Deduplicator.Decorator] for a [message.PublisherDecorator].
+//
+// Use [NewMessageHasherAdler32] for fast tagging or
+// [NewMessageHasherSHA256] for minimal collisions.
+//
+// Window specifies the minimum duration of  how long the
+// duplicate tags are remembered for. Real duration can
+// extend up to 50% longer because it depends on the
+// clean up cycle.
+func NewDeduplicator(
+	tagger MessageHasher,
+	window time.Duration,
+) *Deduplicator {
+	if tagger == nil {
+		panic("cannot use a <nil> tagger")
+	}
+	if window < time.Millisecond {
+		panic("deduplication window of less than a millisecond is impractical")
+	}
+
+	d := &Deduplicator{
+		tagger: tagger,
+		window: window,
+
+		mu:   &sync.Mutex{},
```

Comment:
I wonder if it won't be worth abstracting it with interface - so you could provide it under in-memory implementation.

It has one nice advantage - someone could use redis/whatever for that and handle more use cases.


WDYT?